## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#234c8013-7b1a-4aa7-8524-9914ce155236) |
| mint-gateway-azure.sh | :x: [more...](#7ece43e5-a48c-41af-90a4-78062df62f32) |
| mint-dist-erasure.sh | :x: [more...](#9c3e5aef-4c1f-44fd-94b7-347535b8521b) |
| mint-zoned.sh | :x: [more...](#151aaa1d-da86-4ee1-b0fd-b5184f19b1ef) |
<a name="7ece43e5-a48c-41af-90a4-78062df62f32"></a>
### 11423-cbc44c3/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30891
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fd897474a6bb:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 3 minutes and 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 2 minutes and 54 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 25 minutes and 51 seconds
(8/15) Running minio-dotnet tests ... done in 4 minutes and 33 seconds
(9/15) Running minio-go tests ... FAILED in 3 minutes and 31 seconds
{
  "args": {
    "bucketName": "minio-go-test-vswdn3v9uwyf43vz",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1069,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... FAILED in 8 minutes and 42 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[user metadata]",
  "duration": 638,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = AuthenticationFailed, message = -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=AuthenticationFailed) =====\nDescription=Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.\nRequestId:40e43583-801e-00ea-0c3d-fa12ef000000\nTime:2021-02-03T14:59:42.5922239Z, Details: \n   AuthenticationErrorDetail: The MAC signature found in the HTTP request 'GeFesbi8Crfm/f4rAyDc0VqezWKed8qrscio+cHJNfE=' is not the same as any computed signature. Server used following string to sign: 'PUT\n\n\n128\n\napplication/xml\n\n\n\n\n\n\nx-ms-blob-cache-control:\nx-ms-blob-content-disposition:\nx-ms-blob-content-encoding:\nx-ms-blob-content-language:\nx-ms-blob-content-type:application/octet-stream\nx-ms-client-request-id:61ca5568-5a8f-4350-5ca3-d676d0f732df\nx-ms-date:Wed, 03 Feb 2021 14:59:42 GMT\nx-ms-meta-my_header1:a   b   c\nx-ms-meta-my_header2:\"a   b   c\"\nx-ms-meta-my_project:Project One\nx-ms-meta-my_unicode_tag:å•†å“\nx-ms-version:2019-02-02\n/minioazure/minio-java-test-2n3h244/minio-java-test-2ephpck\ncomp:blocklist\ntimeout:1501'.\n   Code: AuthenticationFailed\n   PUT https://minioazure.blob.core.windows.net/minio-java-test-2n3h244/minio-java-test-2ephpck?comp=blocklist&timeout=1501\n   Authorization: REDACTED\n   Content-Length: [128]\n   Content-Type: [application/xml]\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2021-02-03T14:03:39Z]\n   X-Ms-Blob-Cache-Control: []\n   X-Ms-Blob-Content-Disposition: []\n   X-Ms-Blob-Content-Encoding: []\n   X-Ms-Blob-Content-Language: []\n   X-Ms-Blob-Content-Type: [application/octet-stream]\n   X-Ms-Client-Request-Id: [61ca5568-5a8f-4350-5ca3-d676d0f732df]\n   X-Ms-Date: [Wed, 03 Feb 2021 14:59:42 GMT]\n   X-Ms-Meta-My_header1: [a   b   c]\n   X-Ms-Meta-My_header2: [\"a   b   c\"]\n   X-Ms-Meta-My_project: [Project One]\n   X-Ms-Meta-My_unicode_tag: [商品]\n   X-Ms-Version: [2019-02-02]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.\n   Content-Length: [1092]\n   Content-Type: [application/xml]\n   Date: [Wed, 03 Feb 2021 14:59:42 GMT]\n   Server: [Microsoft-HTTPAPI/2.0]\n   X-Ms-Error-Code: [AuthenticationFailed]\n   X-Ms-Request-Id: [40e43583-801e-00ea-0c3d-fa12ef000000]\n\n\n, bucketName = minio-java-test-2n3h244, objectName = minio-java-test-2ephpck, resource = /minio-java-test-2n3h244/minio-java-test-2ephpck, requestId = 166044DCFA1748BF, hostId = 6755a4dc-6476-4ce8-8090-abb449510d2f)\nrequest={method=PUT, url=http://minio-c2.minio.io:30891/minio-java-test-2n3h244/minio-java-test-2ephpck, headers=x-amz-meta-My-Unicode-Tag: 商品\nx-amz-meta-My-Project: Project One\nx-amz-meta-My-header1: a   b   c\nx-amz-meta-My-Header2: \"a   b   c\"\nContent-Type: application/octet-stream\nHost: minio-c2.minio.io:30891\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: A9oFTxee7YVcJ9fWsgQeKg==\nx-amz-content-sha256: 1ff7959f86334ddc5c188a5083268f600146328b2b6c5185e75bf7d9387d6b74\nx-amz-date: 20210203T145942Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210203/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date;x-amz-meta-my-header1;x-amz-meta-my-header2;x-amz-meta-my-project;x-amz-meta-my-unicode-tag, Signature=*REDACTED*\n}\nresponse={code=403, headers=Accept-Ranges: bytes\nContent-Length: 3086\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 166044DCFA1748BF\nX-Xss-Protection: 1; mode=block\nDate: Wed, 03 Feb 2021 14:59:43 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:890), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(9/15) Running minio-js tests ... FAILED in 1 minutes and 18 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 378,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(9/15) Running minio-py tests ... done in 26 minutes and 26 seconds
(10/15) Running s3cmd tests ... done in 16 minutes and 54 seconds
(11/15) Running s3select tests ... done in 49 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="151aaa1d-da86-4ee1-b0fd-b5184f19b1ef"></a>
### 11423-cbc44c3/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32362
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6d06ae1f6ec3:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-vy9pssen3kdl15",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 252,
  "error": "BucketNotEmpty: The bucket you tried to delete is not empty\n\tstatus code: 409, request id: 1660422671A89B99, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 2 seconds
(2/15) Running aws-sdk-php tests ... done in 44 seconds
(3/15) Running aws-sdk-ruby tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-0a52daac909c"
  },
  "duration": 585.81,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-0a52daac909c', The bucket you tried to delete is not empty",
  "status": "FAIL"
}
(3/15) Running awscli tests ... FAILED in 8 seconds
{
  "name": "awscli",
  "duration": 3614,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-2066/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-2066 An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: The bucket you tried to delete is not empty"
}
(3/15) Running healthcheck tests ... done in 1 seconds
(4/15) Running mc tests ... done in 1 minutes and 2 seconds
(5/15) Running minio-dotnet tests ... done in 42 seconds
(6/15) Running minio-go tests ... FAILED in 2 minutes and 17 seconds
{
  "args": {
    "bucketName": "minio-go-test-olp4btwcn4u4q4nx",
    "objectName": "0vw9c00wwyxay6a33qktwlzxucql62"
  },
  "duration": 493,
  "error": "The bucket you tried to delete is not empty",
  "function": "GetObject(bucketName, objectName)",
  "message": "Cleanup failed",
  "name": "minio-go: testGetObjectReadSeekFunctional",
  "status": "FAIL"
}
(6/15) Running minio-java tests ... FAILED in 13 seconds
(6/15) Running minio-js tests ... FAILED in 25 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 5,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(6/15) Running minio-py tests ... FAILED in 10 minutes and 2 seconds
{
  "name": "minio-py:test_fput_object_small_file",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-5d2cde8c-343e-4430-988c-00360a376b23",
    "object_name": "f4c8c796-0aa4-49ad-b85e-b9d21189cbc3-f",
    "file_path": "/mint/data/datafile-1-MB",
    "metadata": {
      "x-amz-storage-class": "STANDARD_IA"
    }
  },
  "message": "HTTPConnectionPool(host='minio-dev6.minio.io', port=32362): Max retries exceeded with url: /minio-py-test-5d2cde8c-343e-4430-988c-00360a376b23 (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 334, in test_fput_object_small_file\n    _test_fput_object(bucket_name, object_name, _TEST_FILE, metadata, sse)\n  File \"/mint/run/core/minio-py/tests.py\", line 313, in _test_fput_object\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 278, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 836, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 573, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev6.minio.io', port=32362): Max retries exceeded with url: /minio-py-test-5d2cde8c-343e-4430-988c-00360a376b23 (Caused by ResponseError('too many 503 error responses',))\n",
  "duration": 600786
}
(6/15) Running s3cmd tests ... done in 21 seconds
(7/15) Running s3select tests ... FAILED in 5 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-14315c00-7b0e-4b62-a1b2-a28b539c847d"
  },
  "duration": 4986,
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-14315c00-7b0e-4b62-a1b2-a28b539c847d, request_id: 16604306E4FA1697, host_id: 547c04ca-36ef-4964-b84c-9d69fab775e7, bucket_name: s3select-test-14315c00-7b0e-4b62-a1b2-a28b539c847d",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 120, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-14315c00-7b0e-4b62-a1b2-a28b539c847d, request_id: 16604306E4FA1697, host_id: 547c04ca-36ef-4964-b84c-9d69fab775e7, bucket_name: s3select-test-14315c00-7b0e-4b62-a1b2-a28b539c847d\n",
  "status": "FAIL"
}
(7/15) Running security tests ... done in 0 seconds

Executed 7 out of 15 tests successfully.
```
<a name="234c8013-7b1a-4aa7-8524-9914ce155236"></a>
### 11423-cbc44c3/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31605
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9771b10296fa:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-rj1wgs543ptcem",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 235,
  "error": "BucketNotEmpty: The bucket you tried to delete is not empty\n\tstatus code: 409, request id: 166042267D5E9FC5, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 2 seconds
(2/15) Running aws-sdk-php tests ... done in 44 seconds
(3/15) Running aws-sdk-ruby tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-ea4dba234b9c"
  },
  "duration": 462.14,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-ea4dba234b9c', The bucket you tried to delete is not empty",
  "status": "FAIL"
}
(3/15) Running awscli tests ... FAILED in 8 seconds
{
  "name": "awscli",
  "duration": 3966,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-12455/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-12455 An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: The bucket you tried to delete is not empty"
}
(3/15) Running healthcheck tests ... done in 1 seconds
(4/15) Running mc tests ... done in 1 minutes and 4 seconds
(5/15) Running minio-dotnet tests ... done in 41 seconds
(6/15) Running minio-go tests ... FAILED in 1 minutes and 13 seconds
{
  "args": {
    "bucketName": "minio-go-test-crd45t316xco2jyu",
    "objectName": "zr49iphghelpqppre4rpsfy2i4lcdg"
  },
  "duration": 380,
  "error": "The bucket you tried to delete is not empty",
  "function": "GetObject(bucketName, objectName)",
  "message": "Cleanup failed",
  "name": "minio-go: testGetObjectReadSeekFunctional",
  "status": "FAIL"
}
(6/15) Running minio-java tests ... FAILED in 15 seconds
(6/15) Running minio-js tests ... FAILED in 26 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 10,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(6/15) Running minio-py tests ... FAILED in 1 seconds
{
  "name": "minio-py:test_fput_object_small_file",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-bedf280d-c8cb-4887-85b0-77da2dcc60cc",
    "object_name": "23b3f5c4-aa9b-4586-bb25-121e06a4e8c2-f",
    "file_path": "/mint/data/datafile-1-MB",
    "metadata": {
      "x-amz-storage-class": "STANDARD_IA"
    }
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-bedf280d-c8cb-4887-85b0-77da2dcc60cc, request_id: 16604266DE9DC655, host_id: 9456d120-6b49-4431-a915-0ae4b21593c9, bucket_name: minio-py-test-bedf280d-c8cb-4887-85b0-77da2dcc60cc",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 334, in test_fput_object_small_file\n    _test_fput_object(bucket_name, object_name, _TEST_FILE, metadata, sse)\n  File \"/mint/run/core/minio-py/tests.py\", line 313, in _test_fput_object\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-bedf280d-c8cb-4887-85b0-77da2dcc60cc, request_id: 16604266DE9DC655, host_id: 9456d120-6b49-4431-a915-0ae4b21593c9, bucket_name: minio-py-test-bedf280d-c8cb-4887-85b0-77da2dcc60cc\n",
  "duration": 371
}
(6/15) Running s3cmd tests ... done in 30 seconds
(7/15) Running s3select tests ... FAILED in 4 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-1172f15d-fce0-4426-bc07-da424749a540"
  },
  "duration": 3049,
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-1172f15d-fce0-4426-bc07-da424749a540, request_id: 1660426EC323DBA1, host_id: 9456d120-6b49-4431-a915-0ae4b21593c9, bucket_name: s3select-test-1172f15d-fce0-4426-bc07-da424749a540",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 120, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-1172f15d-fce0-4426-bc07-da424749a540, request_id: 1660426EC323DBA1, host_id: 9456d120-6b49-4431-a915-0ae4b21593c9, bucket_name: s3select-test-1172f15d-fce0-4426-bc07-da424749a540\n",
  "status": "FAIL"
}
(7/15) Running security tests ... done in 0 seconds

Executed 7 out of 15 tests successfully.
```
<a name="9c3e5aef-4c1f-44fd-94b7-347535b8521b"></a>
### 11423-cbc44c3/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31537
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d1b40fa99868:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-1jq3quv3aa3cao",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 201,
  "error": "BucketNotEmpty: The bucket you tried to delete is not empty\n\tstatus code: 409, request id: 1660422666805074, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 1 seconds
(2/15) Running aws-sdk-php tests ... done in 44 seconds
(3/15) Running aws-sdk-ruby tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-5f2aadb41b4a"
  },
  "duration": 466.41,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-5f2aadb41b4a', The bucket you tried to delete is not empty",
  "status": "FAIL"
}
(3/15) Running awscli tests ... FAILED in 15 seconds
{
  "name": "awscli",
  "duration": 10925,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-18514/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-18514 An error occurred (BucketNotEmpty) when calling the DeleteBucket operation: The bucket you tried to delete is not empty"
}
(3/15) Running healthcheck tests ... done in 0 seconds
(4/15) Running mc tests ... done in 1 minutes and 6 seconds
(5/15) Running minio-dotnet tests ... done in 39 seconds
(6/15) Running minio-go tests ... FAILED in 1 minutes and 22 seconds
{
  "args": {
    "bucketName": "minio-go-test-uajccmpmf5vbqf3y",
    "objectName": "sg0o9q5ol62fqsqukfx0a4jx23kyiy"
  },
  "duration": 288,
  "error": "The bucket you tried to delete is not empty",
  "function": "GetObject(bucketName, objectName)",
  "message": "Cleanup failed",
  "name": "minio-go: testGetObjectReadSeekFunctional",
  "status": "FAIL"
}
(6/15) Running minio-java tests ... FAILED in 13 seconds
(6/15) Running minio-js tests ... FAILED in 35 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 8,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(6/15) Running minio-py tests ... FAILED in 1 seconds
{
  "name": "minio-py:test_fput_object_small_file",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-d69932ac-9322-4bca-ba24-8757cb9715cc",
    "object_name": "ff5b1114-803c-4ef7-a17d-a279288f206c-f",
    "file_path": "/mint/data/datafile-1-MB",
    "metadata": {
      "x-amz-storage-class": "STANDARD_IA"
    }
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-d69932ac-9322-4bca-ba24-8757cb9715cc, request_id: 1660426BD76FE125, host_id: dae0324a-08db-4779-932a-a3042fc15261, bucket_name: minio-py-test-d69932ac-9322-4bca-ba24-8757cb9715cc",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 334, in test_fput_object_small_file\n    _test_fput_object(bucket_name, object_name, _TEST_FILE, metadata, sse)\n  File \"/mint/run/core/minio-py/tests.py\", line 313, in _test_fput_object\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-d69932ac-9322-4bca-ba24-8757cb9715cc, request_id: 1660426BD76FE125, host_id: dae0324a-08db-4779-932a-a3042fc15261, bucket_name: minio-py-test-d69932ac-9322-4bca-ba24-8757cb9715cc\n",
  "duration": 411
}
(6/15) Running s3cmd tests ... done in 28 seconds
(7/15) Running s3select tests ... FAILED in 3 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-c5e49d4a-1f6d-48e8-81a5-ca482ecdf264"
  },
  "duration": 2955,
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-c5e49d4a-1f6d-48e8-81a5-ca482ecdf264, request_id: 16604272F80AC4FC, host_id: dae0324a-08db-4779-932a-a3042fc15261, bucket_name: s3select-test-c5e49d4a-1f6d-48e8-81a5-ca482ecdf264",
  "error": "Traceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 120, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-c5e49d4a-1f6d-48e8-81a5-ca482ecdf264, request_id: 16604272F80AC4FC, host_id: dae0324a-08db-4779-932a-a3042fc15261, bucket_name: s3select-test-c5e49d4a-1f6d-48e8-81a5-ca482ecdf264\n",
  "status": "FAIL"
}
(7/15) Running security tests ... done in 0 seconds

Executed 7 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/erasure-healing.go
+++ b/cmd/erasure-healing.go
@@ -709,7 +709,7 @@ func (er erasureObjects) purgeObjectDangling(ctx context.Context, bucket, object
 			if versionID == "" {
 				err = er.deleteObject(ctx, bucket, object, writeQuorum)
 			} else {
-				err = er.deleteObjectVersion(ctx, bucket, object, writeQuorum, FileInfo{VersionID: versionID})
+				err = er.deleteObjectVersion(ctx, bucket, object, writeQuorum, FileInfo{VersionID: versionID}, false)
 			}
 
 			// If Delete was successful, make sure to return the appropriate error

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -414,7 +414,7 @@ func (er erasureObjects) getObjectFileInfo(ctx context.Context, bucket, object s
 					er.deleteObjectVersion(ctx, bucket, object, 1, FileInfo{
 						Name:      object,
 						VersionID: opts.VersionID,
-					})
+					}, false)
 				}
 			}
 		}
@@ -800,7 +800,7 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 	return fi.ToObjectInfo(bucket, object), nil
 }
 
-func (er erasureObjects) deleteObjectVersion(ctx context.Context, bucket, object string, writeQuorum int, fi FileInfo) error {
+func (er erasureObjects) deleteObjectVersion(ctx context.Context, bucket, object string, writeQuorum int, fi FileInfo, forceDelMarker bool) error {
 	defer ObjectPathUpdated(pathJoin(bucket, object))
 	disks := er.getDisks()
 	g := errgroup.WithNErrs(len(disks))
@@ -810,7 +810,7 @@ func (er erasureObjects) deleteObjectVersion(ctx context.Context, bucket, object
 			if disks[index] == nil {
 				return errDiskNotFound
 			}
-			return disks[index].DeleteVersion(ctx, bucket, object, fi)
+			return disks[index].DeleteVersion(ctx, bucket, object, fi, forceDelMarker)
 		}, index)
 	}
 	// return errors if any during deletion
@@ -1098,7 +1098,7 @@ func (er erasureObjects) DeleteObject(ctx context.Context, bucket, object string
 			// version as delete marker
 			// Add delete marker, since we don't have any version specified explicitly.
 			// Or if a particular version id needs to be replicated.
-			if err = er.deleteObjectVersion(ctx, bucket, object, writeQuorum, fi); err != nil {
+			if err = er.deleteObjectVersion(ctx, bucket, object, writeQuorum, fi, opts.DeleteMarker); err != nil {
 				return objInfo, toObjectErr(err, bucket, object)
 			}
 			return fi.ToObjectInfo(bucket, object), nil
@@ -1115,7 +1115,7 @@ func (er erasureObjects) DeleteObject(ctx context.Context, bucket, object string
 		DeleteMarkerReplicationStatus: opts.DeleteMarkerReplicationStatus,
 		VersionPurgeStatus:            opts.VersionPurgeStatus,
 		TransitionStatus:              opts.TransitionStatus,
-	}); err != nil {
+	}, opts.DeleteMarker); err != nil {
 		return objInfo, toObjectErr(err, bucket, object)
 	}
 

--- a/cmd/naughty-disk_test.go
+++ b/cmd/naughty-disk_test.go
@@ -245,11 +245,11 @@ func (d *naughtyDisk) WriteMetadata(ctx context.Context, volume, path string, fi
 	return d.disk.WriteMetadata(ctx, volume, path, fi)
 }
 
-func (d *naughtyDisk) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo) (err error) {
+func (d *naughtyDisk) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo, forceDelMarker bool) (err error) {
 	if err := d.calcError(); err != nil {
 		return err
 	}
-	return d.disk.DeleteVersion(ctx, volume, path, fi)
+	return d.disk.DeleteVersion(ctx, volume, path, fi, forceDelMarker)
 }
 
 func (d *naughtyDisk) ReadVersion(ctx context.Context, volume, path, versionID string, readData bool) (fi FileInfo, err error) {

--- a/cmd/storage-interface.go
+++ b/cmd/storage-interface.go
@@ -55,7 +55,7 @@ type StorageAPI interface {
 	WalkVersions(ctx context.Context, volume, dirPath, marker string, recursive bool, endWalkCh <-chan struct{}) (chan FileInfoVersions, error)
 
 	// Metadata operations
-	DeleteVersion(ctx context.Context, volume, path string, fi FileInfo) error
+	DeleteVersion(ctx context.Context, volume, path string, fi FileInfo, forceDelMarker bool) error
 	DeleteVersions(ctx context.Context, volume string, versions []FileInfo) []error
 	WriteMetadata(ctx context.Context, volume, path string, fi FileInfo) error
 	ReadVersion(ctx context.Context, volume, path, versionID string, readData bool) (FileInfo, error)

--- a/cmd/storage-rest-client.go
+++ b/cmd/storage-rest-client.go
@@ -332,10 +332,11 @@ func (client *storageRESTClient) WriteMetadata(ctx context.Context, volume, path
 	return err
 }
 
-func (client *storageRESTClient) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo) error {
+func (client *storageRESTClient) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo, forceDelMarker bool) error {
 	values := make(url.Values)
 	values.Set(storageRESTVolume, volume)
 	values.Set(storageRESTFilePath, path)
+	values.Set(storageRESTForceDelMarker, strconv.FormatBool(forceDelMarker))
 
 	var buffer bytes.Buffer
 	if err := msgp.Encode(&buffer, &fi); err != nil {

--- a/cmd/storage-rest-common.go
+++ b/cmd/storage-rest-common.go
@@ -17,7 +17,7 @@
 package cmd
 
 const (
-	storageRESTVersion       = "v26" // Add NumVersions/SuccessorModTime fields in FileInfo
+	storageRESTVersion       = "v27" // Add force-delete-marker to DeleteVersion
 	storageRESTVersionPrefix = SlashSeparator + storageRESTVersion
 	storageRESTPrefix        = minioReservedBucketPath + "/storage"
 )
@@ -58,6 +58,7 @@ const (
 	storageRESTVolumes        = "volumes"
 	storageRESTDirPath        = "dir-path"
 	storageRESTFilePath       = "file-path"
+	storageRESTForceDelMarker = "force-delete-marker"
 	storageRESTVersionID      = "version-id"
 	storageRESTReadData       = "read-data"
 	storageRESTTotalVersions  = "total-versions"

--- a/cmd/storage-rest-server.go
+++ b/cmd/storage-rest-server.go
@@ -300,6 +300,11 @@ func (s *storageRESTServer) DeleteVersionHandler(w http.ResponseWriter, r *http.
 	vars := mux.Vars(r)
 	volume := vars[storageRESTVolume]
 	filePath := vars[storageRESTFilePath]
+	forceDelMarker, err := strconv.ParseBool(vars[storageRESTForceDelMarker])
+	if err != nil {
+		s.writeErrorResponse(w, errInvalidArgument)
+		return
+	}
 
 	if r.ContentLength < 0 {
 		s.writeErrorResponse(w, errInvalidArgument)
@@ -312,7 +317,7 @@ func (s *storageRESTServer) DeleteVersionHandler(w http.ResponseWriter, r *http.
 		return
 	}
 
-	err := s.storage.DeleteVersion(r.Context(), volume, filePath, fi)
+	err = s.storage.DeleteVersion(r.Context(), volume, filePath, fi, forceDelMarker)
 	if err != nil {
 		s.writeErrorResponse(w, err)
 	}
@@ -1038,7 +1043,7 @@ func registerStorageRESTHandlers(router *mux.Router, endpointServerPools Endpoin
 			subrouter.Methods(http.MethodPost).Path(storageRESTVersionPrefix + storageRESTMethodWriteMetadata).HandlerFunc(httpTraceHdrs(server.WriteMetadataHandler)).
 				Queries(restQueries(storageRESTVolume, storageRESTFilePath)...)
 			subrouter.Methods(http.MethodPost).Path(storageRESTVersionPrefix + storageRESTMethodDeleteVersion).HandlerFunc(httpTraceHdrs(server.DeleteVersionHandler)).
-				Queries(restQueries(storageRESTVolume, storageRESTFilePath)...)
+				Queries(restQueries(storageRESTVolume, storageRESTFilePath, storageRESTForceDelMarker)...)
 			subrouter.Methods(http.MethodPost).Path(storageRESTVersionPrefix + storageRESTMethodReadVersion).HandlerFunc(httpTraceHdrs(server.ReadVersionHandler)).
 				Queries(restQueries(storageRESTVolume, storageRESTFilePath, storageRESTVersionID, storageRESTReadData)...)
 			subrouter.Methods(http.MethodPost).Path(storageRESTVersionPrefix + storageRESTMethodRenameData).HandlerFunc(httpTraceHdrs(server.RenameDataHandler)).

--- a/cmd/xl-storage-disk-id-check.go
+++ b/cmd/xl-storage-disk-id-check.go
@@ -256,12 +256,12 @@ func (p *xlStorageDiskIDCheck) WriteAll(ctx context.Context, volume string, path
 	return p.storage.WriteAll(ctx, volume, path, b)
 }
 
-func (p *xlStorageDiskIDCheck) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo) (err error) {
+func (p *xlStorageDiskIDCheck) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo, forceDelMarker bool) (err error) {
 	if err = p.checkDiskStale(); err != nil {
 		return err
 	}
 
-	return p.storage.DeleteVersion(ctx, volume, path, fi)
+	return p.storage.DeleteVersion(ctx, volume, path, fi, forceDelMarker)
 }
 
 func (p *xlStorageDiskIDCheck) WriteMetadata(ctx context.Context, volume, path string, fi FileInfo) (err error) {

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -878,7 +878,7 @@ func (s *xlStorage) ListDir(ctx context.Context, volume, dirPath string, count i
 func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions []FileInfo) []error {
 	errs := make([]error, len(versions))
 	for i, version := range versions {
-		if err := s.DeleteVersion(ctx, volume, version.Name, version); err != nil {
+		if err := s.DeleteVersion(ctx, volume, version.Name, version, false); err != nil {
 			errs[i] = err
 		}
 	}
@@ -886,9 +886,9 @@ func (s *xlStorage) DeleteVersions(ctx context.Context, volume string, versions 
 	return errs
 }
 
-// DeleteVersion - deletes FileInfo metadata for path at `xl.meta`. Create a fresh
-// `xl.meta` if it does not exist and we creating a new delete-marker.
-func (s *xlStorage) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo) error {
+// DeleteVersion - deletes FileInfo metadata for path at `xl.meta`. forceDelMarker
+// will force creating a new `xl.meta` to create a new delete marker
+func (s *xlStorage) DeleteVersion(ctx context.Context, volume, path string, fi FileInfo, forceDelMarker bool) error {
 	if HasSuffix(path, SlashSeparator) {
 		return s.Delete(ctx, volume, path, false)
 	}
@@ -898,7 +898,7 @@ func (s *xlStorage) DeleteVersion(ctx context.Context, volume, path string, fi F
 		if err != errFileNotFound {
 			return err
 		}
-		if fi.Deleted {
+		if fi.Deleted && forceDelMarker {
 			// Create a new xl.meta with a delete marker in it
 			return s.WriteMetadata(ctx, volume, path, fi)
 		}

